src/uffd_handler/src/handler.rs

```diff
@@ -1,59 +1,159 @@
-        revents: 0,
-    };
-    // Loop, handling incoming events on the userfaultfd file descriptor.
-    loop {
-        // See what poll() tells us about the userfaultfd.
-        let nready = unsafe { libc::poll(&mut pollfd, 1, -1) };
+        // Make sure memory size matches backing data size.
+        assert_eq!(memsize, size);
 
-        if nready == -1 {
-            panic!("Could not poll for events!")
+        // Get credentials of Firecracker process sent through the stream.
+        let creds: libc::ucred = get_peer_process_credentials(stream);
+
+        Self {
+            mem_regions,
+            backing_buffer: data,
+            uffd,
+            _firecracker_pid: creds.pid as u32,
         }
+    }
 
-        let revents = pollfd.revents;
+    pub fn update_mem_state_mappings(&mut self, start: u64, end: u64, state: &MemPageState) {
+        for region in self.mem_regions.iter_mut() {
+            for (key, value) in region.page_states.iter_mut() {
+                if key >= &start && key < &end {
+                    *value = state.clone();
+                }
+            }
+        }
+    }
 
-        println!(
-            "poll() returns: nready = {}; POLLIN = {}; POLLERR = {}",
-            nready,
-            revents & libc::POLLIN,
-            revents & libc::POLLERR,
-        );
+    fn populate_from_file(&self, region: &MemRegion) -> (u64, u64) {
+        let src = self.backing_buffer as u64 + region.mapping.offset;
+        let start_addr = region.mapping.base_host_virt_addr;
+        let len = region.mapping.size;
+        // Populate whole region from backing mem-file.
+        // This offers an example of how memory can be loaded in RAM,
+        // however this can be adjusted to accommodate use case needs.
+        // SAFETY: this is safe because the parameters are controlled by us.
+        let ret = unsafe {
+            self.uffd
+                .copy(src as *const _, start_addr as *mut _, len, true)
+                .expect("Uffd copy failed")
+        };
 
-        // Read an event from the userfaultfd.
-        let event = uffd_handler
-            .uffd
-            .read_event()
-            .expect("Failed to read uffd_msg")
-            .expect("uffd_msg not ready");
-
-        // We expect to receive either a Page Fault or Removed
-        // event (if the balloon device is enabled).
-        match event {
-            userfaultfd::Event::Pagefault { addr, .. } => uffd_handler.serve_pf(addr as *mut u8),
-            userfaultfd::Event::Remove { start, end } => uffd_handler.update_mem_state_mappings(
-                start as *mut u8 as u64,
-                end as *mut u8 as u64,
-                &MemPageState::Removed,
-            ),
-            _ => panic!("Unexpected event on userfaultfd"),
+        // Make sure the UFFD copied some bytes.
+        assert!(ret > 0);
+
+        (start_addr, start_addr + len as u64)
+    }
+
+    fn zero_out(&mut self, addr: u64) -> (u64, u64) {
```

Comment:
same as above, can we represent the address as an actual pointer (and if not, can we represent the address using `usize`, which I think would be the more appropriate numeric type)?

Reply:
a 32-bit hypervisor supporting a 64-bit sounds... wild, haha.

As mentioned above, am happy with the way things are now, as we have just one pointer type (`* c_void`) and one numeric type (`usize`) to deal with addresses